pep-9999.rst

```diff
@@ -0,0 +1,669 @@
+list is not immutable. Thus, the ``print`` function in the
+newly spawned thread will be racing with the deletion. In
+Python 3.12, this is not a problem as the GIL prevents this
+race. To ensure `container thread-safety
+<https://peps.python.org/pep-0703/#container-thread-safety>`_,
+PEP 703 proposes per-object locks instead. If ``pair`` was
+frozen, the deletion would have caused an error.
+
+The following image illustrates that as soon as an object *a*
+is reachable by two threads, then all other objects that
+*a* can reach are also reachable by both threads. The dashed
+red references to *c* and *d* are not possible because then
+*c* and *d* would not be in areas where only a single thread
+could reach them.
+
+To map the code example above to the figure -- ``pair`` is *a* and ``list`` is *b*.
+
+.. image:: pep-9999/sharing1.png
+   :width: 50%
+
+See also the discussion about extensions further down in this
+document.
+
+Deep immutability can be implemented efficiently. We discuss
+a read-barrier based approach in the alternatives.
+
+
+Optimisations and Caching Benefits
+----------------------------------
+
+Immutable objects provide opportunities for optimisation, such as
+structural sharing, memoization, and just-in-time (JIT)
+compilation techniques (specialising for immutable data, e.g.
+fixed shape, fewer barriers, inlining, etc.). Freezing objects can
+allow Python to implement more efficient caching mechanisms and
+enable compiler optimisations that rely on immutability
+assumptions. This PEP will permit such opportunities to go
+beyond today's immutable objects (like ``int``, ``string``) and
+*shallow* immutable objects (``tuple``, ``frozenset``).
+
+
+
+Specification
+=============
+
+Changes to Python objects
+-------------------------
+
+Every Python object will have a flag that keeps track of its
+immutability status. Details about the default value of
+this flag is discussed further down in this document. 
+
+The flag can be added without extending the size of the
+Python object header.
+
+
+Implementation of Immutability
+------------------------------
+
+Immutability is enforced through run-time checking. The macro
+``Py_CHECKWRITE(op)`` is inserted on all paths that are guaranteed
+to end up in a write to ``op``. The macro inspects the immutability
+flag in the header of ``op`` and signals an error if the immutability
+flag is set.
+
+A typical use of this check looks like this:
+
+.. code-block:: c
+
+	if (!Py_CHECKWRITE(op)) {        // perform the check
+	    PyErr_WriteToImmutable(op);  // raise the error if the check fails
+	    return NULL;                 // abort the write
+	}  
+	... // code that performs the write
+
+
+As writes are common but lack a common path that most writes to through
+the PEP requires a ``Py_CHECKWRITE`` call, there are several places in
+the CPython code base that are changed as a consequence of this PEP.
+So far we have identified around 70 writes spread across a dozen files.
```

Comment:
Would these `Py_CHECKWRITE` checks be added mainly at the public C API boundary, or are there places where they are required in internal functions as well.  The surprisingly low number of 70 makes me think it's the former, but maybe clarifying that here would be helpful.

Things like 

```
assert(Py_CHECKWRITE(op))
```

in a number of internal places would probably also be helpful for debugging.